### PR TITLE
Support for "op_return" parameters in bitcoincash: style URI's on Qt …

### PR DESCRIPTION
**Pull request by jimtendo**
…client

### CODE REVIEW

**Comment by cculianu** [comment]
I'll have to review this -- I don't think the OP_RETURN message is always read if the config variable in self.config for op_return is set to false.  Will have to double-check it.  Otherwise this looks good -- although is there a standard naming for this variable yet? Like some official bitcoincash: uri guideline? Or?

**Comment by jimtendo** [comment]
You are correct actually. On line 1453 we have:

`opreturn_message = self.message_opreturn_e.text() if self.config.get('enable_opreturn') else None`

Changing this to just:

`opreturn_message = self.message_opreturn_e.text()`

... appears to work, but I'm not sure if there's some security implication I'm missing (this setting is disabled in EC by default). Can you think of a reason why this check would need to be performed as opposed to just checking whether that field has length?

As regards to the standard naming, there is nothing yet. I did create a few unofficial posts to get community feedback but could not get any suggestions on this front.

**Comment by markblundeberg** [comment]
Interesting idea! Would be good to standardize this with other people tho...

**Comment by jimtendo** [comment]
@markblundeberg agree on the standardization. I'm not sure how to best go about doing this though. Previously, I had tried to get feedback via Reddit, but was not getting much traction. Here is the thread describing a few use cases for this:

https://www.reddit.com/r/btc/comments/9w0n59/walleturi_support_for_op_return/

**Comment by markblundeberg** [comment]
I would like an OP_RETURN functionality in Electron Cash send dialog (and in the URI) that supports encoding of arbitrary binary data. Is this possible? I know URIs have percent-encoding but I'm not sure if arbitrary bytestrings are allowed, and likewise it would mean we should settle on an encoding scheme for the Op_Return input field in the send dialog.

@jscramer made a pseudo-markup language for the Op_Return field, so that you could write `<hex>ffffff` but maybe this is not the optimal way.

There is one limitation with this op_return thing URI right now, which is that it is limited to a single PUSH in script, after OP_RETURN. But I think this is totally fine.

**Comment by cculianu** [comment]
> ou are correct actually. On line 1453 we have:
> 
> `opreturn_message = self.message_opreturn_e.text() if self.config.get('enable_opreturn') else None`
> 
> Changing this to just:
> 
> `opreturn_message = self.message_opreturn_e.text()`
> 

Hmm. That's one approach but the OP_RETURN does increase the TX size.  If they really enabled OP_RETURN in their conf it would make sense to at least SHOW them the OP_RETURN field and prompt them or something -- not just go behind the user's back about it because they clicked a URL.

I'll review this and let you know my thoughts when I get time later today or tomorrow -- thanks a lot for this work and it's a good idea!

**Comment by jimtendo** [comment]
@cculianu we are actually showing this field if op_return is present in the URL (doing a setHidden = false in code).

I understand why this is hidden by default, particularly given BTC's heavy tx fees, but I think for BCH so long as we show that the field is being set (and give the user the option to blank it), personally, I don't think there should be too much of concern here.

I'd still like to get other's thoughts on the naming of the parameter and whether op_return (snake-case) is really the best to use.

**Comment by markblundeberg** [comment]
`op_return` is absolutely the right name, fight me if you disagree. :-D

(Really though, it's OP_RETURN everywhere else so camel case would be wrong. Snake case is good.)

**Comment by cculianu** [comment]
> @cculianu we are actually showing this field if op_return is present in the URL (doing a setHidden = false in code).

@jimtendo I get what you're saying but for "historical/hysterical reasons" the Send tab can optionally be kept "pure" and free of clutter hence no OP_RETURN field.

Can you modify your PR to, in do_clear(), undo the damage from setHidden(False) if they have op_return disabled in the conf? (it could be that do_clear() already undoes the damage unconditionally.. haven't checked.)

We should respect the user's wishes about non-cluttered UIs and that conf option was put in place by the mind of Jonald for a reason.. :)

@markblundeberg I agree.. that's a good name.  Let's stick to it.

**Comment by cculianu** [comment]
> I would like an OP_RETURN functionality in Electron Cash send dialog (and in the URI) that supports encoding of arbitrary binary data. Is this possible? I know URIs have percent-encoding but I'm not sure if arbitrary bytestrings are allowed, and likewise it would mean we should settle on an encoding scheme for the Op_Return input field in the send dialog.
> 
> @jscramer made a pseudo-markup language for the Op_Return field, so that you could write `<hex>ffffff` but maybe this is not the optimal way.
> 

Hmm this is a very good idea.  I am ok with any encoding scheme.  <hex>whatever</hex> works... Although I can't help but think perhaps it would create bugs for nubie programmers to include < and > symbols in a URL.  Another option is have 2 different vars: op_return and op_return_raw and the raw variant uses some simple format like a string of hex-encoded bytes ex `ff6d1397...`

Just a thought.

File: gui/qt/main_window.py [hunk 1]
```diff
@@ -1722,6 +1726,8 @@ def do_clear(self):
         self.max_button.setDisabled(False)
         self.set_pay_from([])
         self.tx_external_keypairs = {}
+        self.message_opreturn_e.setVisible(self.config.get('enable_opreturn'))
```
Comment (cculianu) [inline gui/qt/main_window.py]: 👍

**Comment by cculianu** [comment]
Looks good.  Only "gotcha" is the tx generation code with respect to the conf options -- I want to make sure it's not possible through pathological uses of the UI for this to not behave as expected (eg old OP_RETURN strings left in the UI invisible or if conf option is off, OP_RETURN not respected even if shown in UI due to URL).

Let me verify  .. otherwise this looks good.  Thanks for the change @jimtendo ..

**Comment by jimtendo** [comment]
@cculianu I did check this briefly... it looks like it'll be cleared when the enable_op_return setting is flicked.

```
    def on_opret(x):
            self.config.set_key('enable_opreturn', bool(x))
            if not x:
                self.message_opreturn_e.setText("")
                self.op_return_toolong = False
            self.message_opreturn_e.setHidden(not x)
            self.opreturn_label.setHidden(not x)
```

**Comment by cculianu** [comment]
ok, that looks good.

1 more bit of paranoia. Since we are modifying previous assumptions about the UI state here with this change -- we need to be careful that the UX doesn't end up being crappy in pathological circumstances:

- What happens if user clicks 2 URIs in rapid succession (without them actually sending the tx).  EG if they (1) have OP_RETURN **disabled** in conf, and (2) click the first URI  which **has** an op_return variable and then (3) click another URI (without sending the first tx) that does **not** have op_return?  Ideal behavior is that the UI resets itself on the second URI to its default state (OP_RETURN field hidden). Verify this is the case if you can by either reading the code or trying it.

- To enforce this -- Can you verify that the parse_URI function call-path ends up calling do_clear() before it begins setting up the form?  It should be calling do_clear().. but it if doesn't -- add a call to it in there wherever it makes sense to do so.

The idea is just to make sure the UX can't produce inconsistent/confusing/broken-looking states for users in pathological circumstances... Since this change does cause previous invariants to no longer be invariant. :)

**Comment by cculianu** [comment]
@jimtendo Good good good.  I'm glad you caught that unconditionally clearing it could create problems. True. 

Your if/elif block looks fine.  Don't worry -- it's "sufficiently pythonic".  And I think it's the correct way to handle it. :)

So.. put it in another commit and .. we can merge it soon!

**Comment by cculianu** [comment]
well.. this looks good.  I'm merging it.  What have we got to lose? The world is on fire today anyway! This looks good to me!

**Comment by cculianu** [comment]
@jimtendo thanks for this! 👍

**Comment by cculianu** [comment]
TBH, as I play with this feature I like it more and more.  thanks again!

**Comment by jimtendo** [comment]
No problem, thank you very much for merging. I'd like to get other wallets to support this functionality also as quite a few services I'm working on would need to leverage this to work correctly.

The primary use-case from my end is being able to track specific transactions made through services I plan to run without having to act as an intermediary for funds. It eliminates either:

- the user having to deposit funds into a site-run wallet that supports this functionality
- the service having to act as a proxy of funds (i.e. first send to a service-controlled address which forwards to the recipient)

Instead, to identify relevant transactions, I can just specify an OP_RETURN parameter to attach to the transaction and the users can then use their native wallet.

P2P money ;)

**Comment by markblundeberg** [comment]
@jimtendo we recently (#1086 ) added a second type of op_return user input field where the user can provide the raw hex script that appears after the OP_RETURN opcode. The hex stuff includes everthing: push opcodes, etc.. This gives the ultimate freedom over multi-push op_returns.

If we were to add this to the uri scheme, what would you prefer as a name? Something like `op_return_raw=047370616d` ?

**Comment by jimtendo** [comment]
@markblundeberg regarding URI with both `op_return` and `op_return_raw` my thoughts are that `op_return` should take precedence. My reasoning behind this is that I think if both params existed, it's likely the one constructing the URI is doing something malicious (e.g. the `op_return` might be inconspicuous text, but a sublte `op_return_raw` is thrown in there that includes a tracking ID). User could miss this when hovering over a URL and sending the transaction. Curious on other's thoughts here though.

As to the empty op_return output - personally, I cannot see a use case, but that does not mean there isn't one.

Just FYI, my use-case was to identify particular transactions and attach them to a user on a platform - without having to create/track a mediating address for the transaction. Unfortunately, work has been keeping me too busy to work on this lately, but I hope to have a Beta available sometime next month.

**Pull request merged**